src/Npgsql/NpgsqlDataSource.cs

```diff
@@ -397,34 +397,46 @@ internal virtual bool TryRemovePendingEnlistedConnector(NpgsqlConnector connecto
     #region Dispose
 
     /// <inheritdoc />
-    protected override void Dispose(bool disposing)
+    protected sealed override void Dispose(bool disposing)
     {
         if (disposing && Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
+            DisposeBase();
+    }
+
+    /// <inheritdoc cref="Dispose" />
+    protected virtual void DisposeBase()
+    {
+        var cancellationTokenSource = _timerPasswordProviderCancellationTokenSource;
+        if (cancellationTokenSource is not null)
         {
-            var cancellationTokenSource = _timerPasswordProviderCancellationTokenSource;
-            if (cancellationTokenSource is not null)
-            {
-                cancellationTokenSource.Cancel();
-                cancellationTokenSource.Dispose();
-            }
+            cancellationTokenSource.Cancel();
+            cancellationTokenSource.Dispose();
+        }
 
-            _passwordProviderTimer?.Dispose();
+        _passwordProviderTimer?.Dispose();
 
-            _setupMappingsSemaphore.Dispose();
+        _setupMappingsSemaphore.Dispose();
 
-            Clear();
-        }
+        Clear();
```

Comment:
We should think about this - I'm not sure it's right to immediately clear all idle connector on Dispose... Imagine there's a big backlog of queued multiplexing commands; Dispose would clear all idle connectors, although we still need them to process the backlog.

Conceptually, I think Dispose should mean "don't accept any more new commands" (i.e. execute starts throwing). But we can defer the actual clearing to the point where we've finished processing the queue, i.e. when a connector gets ChannelClosedException on their in-flight channel, they know there are no more commands coming and should terminate at that point...

What do you think?

Reply:
I tend to agree.  In the current state of https://github.com/npgsql/npgsql/pull/4722 I'm disposing the DataSource as a reaction to Clear[All](). This could cause quite a bit of ugly user experience when queries from unrelated threads get cancelled.

Reply:
@Brar I don't think we're discussing cancelling any queries in any case - at least not at the moment. My comment above was more about whether we should immediately clear the pool (i.e. close idle connectors) when Dispose is called, or wait until the command queue is drained first.

Reply:
Ah, ok. So the side effect would possibly be that it may take even longer until an over-full command queue is drained upon disposal because we're starting to remove the connectors as they become idle?

Reply:
Yeah, something like that. In fact, if we Clear on Dispose and there's still commands pending, we'd even need to open new physical connections, since none are available for executing the pending commands...

Reply:
> My comment above was more about whether we should immediately clear the pool (i.e. close idle connectors) when Dispose is called, or wait until the command queue is drained first.

It should be easy to fix: just put `await base.DisposeAsyncBase` in the multiplexing override last.

Reply:
That's a really nice way to do it @vonzshik.